rust/src/message/traits.rs

```diff
@@ -0,0 +1,400 @@
+use super::DecodeError;
+use crate::{crypto::ByteObject, EncrMaskSeed, LocalSeedDict, SumParticipantPublicKey};
+use anyhow::{anyhow, Context};
+use std::{
+    convert::{TryFrom, TryInto},
+    io::{Cursor, Write},
+    ops::Range,
+};
+
+/// `ToBytes` is implemented by types that can be serialized
+pub trait ToBytes {
+    /// Length of the buffer for encoding the type
+    fn buffer_length(&self) -> usize;
+    /// Serialize the type in the given buffer.
+    ///
+    /// # Panic
+    ///
+    /// This method may panic if the given buffer is too small. Thus,
+    /// [`buffer_length`] must be called prior to calling `to_bytes`,
+    /// and a large enough buffer must be provided.
+    fn to_bytes<T: AsMut<[u8]>>(&self, buffer: &mut T);
+}
+
+/// `FromBytes` is the counterpart of `ToBytes`. It is implemented by
+/// types that can be deserialized.
+pub trait FromBytes: Sized {
```

Comment:
Maybe it is better to call it `TryFromBytes` to indicate that it can fail (similar to the `TryFrom` trait).